src/Covenant/Analysis/Poetry/PoetryLicenseParser.cs

```diff
@@ -0,0 +1,120 @@
+using System.Text.RegularExpressions;
+
+namespace Covenant.Analysis.Poetry;
+
+internal static class PoetryLicenseParser
+{
+    public static BomLicense? Parse(string sitePackagesPath, PoetryLockPackage package, AnalysisContext context)
+    {
+        var packageName = package.Name!.Replace("-", "_");
+        var jsonMetadataPath = System.IO.Path.Join(sitePackagesPath, $"{packageName}-{package.Version}.dist-info", "metadata.json");
+        var metadataPath = System.IO.Path.Join(sitePackagesPath, $"{packageName}-{package.Version}.dist-info", "METADATA");
+        var licensePath = System.IO.Path.Join(sitePackagesPath, $"{packageName}-{package.Version}.dist-info", "LICENSE");
+
+        List<(string Path, Func<string, string?> Parser)> metadataTypes = [
+            (jsonMetadataPath, ParseJsonMetadataFile),
+            (metadataPath, ParseMetadataFile),
+            (licensePath, ParseLicenseFile)
+        ];
+
+        string? licenseString = null;
+
+        foreach (var m in metadataTypes)
+        {
+            if (!File.Exists(m.Path))
+            {
+                continue;
+            }
+
+            licenseString = m.Parser(m.Path);
+
+            if (licenseString != null)
+            {
+                break;
+            }
+        }
+
+        if (licenseString == null)
+        {
+            context.AddWarning($"Could not find license for package [yellow]{package.Name}[/] ({package.Version}). Tried metadata.json, METADATA and LICENSE files.");
+        }
+
+        return Parse(licenseString);
+    }
+
+    private static string? ParseLicenseFile(string licensePath)
+    {
+        return File.ReadAllText(licensePath);
+    }
+
+    private static string? ParseJsonMetadataFile(string packageMetadataPath)
+    {
+        using (var packageMetadataReader = new StreamReader(packageMetadataPath))
+        {
+            var metadata = JsonConvert.DeserializeObject<Dictionary<string, object>>(packageMetadataReader.ReadToEnd());
+            return metadata?["license"]?.ToString();
```

Comment:
Again this looks slightly like "keep passing the nullability buck until the compiler stops complaining about this particular code".

The first `?` on this line essentially says that you sometimes expect the preceding call to `JsonConver.DeserializeObect` to return `null`. That will only happen if the file at `packageMetdataPath` contains literally the text `null`.

The only reason `DeserializeObect` is declared as returning `null` is because they have to deal with that possibility. The text "null" is, technically, valid JSON. So if the file contains just those four characters, it can't reject it because it's properly formed JSON, and so it returns a .NET `null` to represent that JSON `null`.

But it's somewhat rare for that to be non-exceptional. If you would consider it to be a mistake for the file here to contain just the text `null`, then it would be appropriate to add something like this:

```cs
if (metadata is null)
{
    throw new ArgumentException("JSON file contains just the text 'null' when an object was expected");
}
```

If you add that before the `return` statement you'll find you can remove the first of the `?` marks because the compiler will know the code can't reach that statement unless `metadata` is non-null. So it would become this:

```cs
return metadata["license"]?.ToString();
```

So now let's talk about that second `?`. Is it expected that sometimes there will be no `license` property in the JSON? If so, then I guess `string?` is a suitable return type, and you do need that `?` in there. But if it's always a mistake for that file not to have a `license`, and you don't expect to see that in normal usage, consider an exception in that case (at which point you can remove the `?` marks on this line and also the return type of the method, and that may in turn remove the need to add a bunch of null handling in code that calls this.)